File: stealthpilot-ai-interview-assistant-electron/electron/groqService.js
/**
 * Groq / Gemma Response Generation Service
 *
 * Primary: Groq API with daily model rotation and streaming SSE
 * Fallback: Gemma-3-27b-it via Google GenAI SDK
 *
 * Mirrors the cheating-daddy approach — Gemini Live handles audio transcription,
 * then we send the transcribed text to Groq (or Gemma) for a fast text response.
 */

const { BrowserWindow } = require('electron');
const Store = require('electron-store');

const store = new Store();

// Conversation history for multi-turn context
let groqConversationHistory = [];
let currentSystemPrompt = 'You are a helpful assistant.';

/* ── helpers ───────────────────────────────────────────────────────────────── */

function sendToRenderer(channel, data) {
    const windows = BrowserWindow.getAllWindows();
    if (windows.length > 0) {
        windows[0].webContents.send(channel, data);
    }
}

function stripThinkingTags(text) {
    return text.replace(/<think>[\s\S]*?<\/think>/g, '').trim();
}

/* ── Groq daily model rotation ────────────────────────────────────────────── */

const GROQ_MODELS = [
    'qwen-qwq-32b',
    'deepseek-r1-distill-llama-70b',
    'llama-3.3-70b-versatile',
    'llama-3.1-8b-instant',
    'meta-llama/llama-4-scout-17b-16e-instruct',
];

function getModelForToday() {
    const dayOfYear = Math.floor(
        (Date.now() - new Date(new Date().getFullYear(), 0, 0).getTime()) / 86400000
    );
    return GROQ_MODELS[dayOfYear % GROQ_MODELS.length];
}

function getGroqApiKey() {
    return store.get('groqApiKey', '') || '';
}

function getGeminiApiKey() {
    return store.get('geminiApiKey', '') || '';
}

function hasGroqKey() {
    const key = getGroqApiKey();
    return key && key.trim() !== '';
}

/* ── Groq streaming SSE ───────────────────────────────────────────────────── */

async function sendToGroq(transcription) {
    const groqApiKey = getGroqApiKey();
    if (!groqApiKey) {
        console.log('[Groq] No API key, skipping');
        return;
    }
    if (!transcription || transcription.trim() === '') return;

    const modelToUse = getModelForToday();
    console.log(`[Groq] Sending to ${modelToUse}:`, transcription.substring(0, 100) + '...');

    groqConversationHistory.push({ role: 'user', content: transcription.trim() });
    if (groqConversationHistory.length > 20) {
        groqConversationHistory = groqConversationHistory.slice(-20);
    }

    try {
        const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${groqApiKey}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                model: modelToUse,
                messages: [
                    { role: 'system', content: currentSystemPrompt },
                    ...groqConversationHistory,
                ],
                stream: true,
                temperature: 0.7,
                max_tokens: 1024,
            }),
        });

        if (!response.ok) {
            const errText = await response.text();
            console.error('[Groq] API error:', response.status, errText);
            sendToRenderer('update-status', `Groq error: ${response.status}`);
            return;
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let fullText = '';
        let isFirst = true;

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            const chunk = decoder.decode(value, { stream: true });
            const lines = chunk.split('\n').filter((l) => l.trim() !== '');

            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    const data = line.slice(6);
                    if (data === '[DONE]') continue;
                    try {
                        const json = JSON.parse(data);
                        const token = json.choices?.[0]?.delta?.content || '';
                        if (token) {
                            fullText += token;
                            const display = stripThinkingTags(fullText);
                            if (display) {
                                sendToRenderer(isFirst ? 'new-response' : 'update-response', display);
                                isFirst = false;
                            }
                        }
                    } catch (_) {
                        /* skip invalid JSON chunks */
                    }
                }
            }
        }

        const cleaned = stripThinkingTags(fullText);
        if (cleaned) {
            groqConversationHistory.push({ role: 'assistant', content: cleaned });
        }

        // Notify renderer of completed response for session storage
        sendToRenderer('response-complete', {
            transcription: transcription.trim(),
            response: cleaned,
            model: modelToUse,
            provider: 'groq',
        });

        console.log(`[Groq] Response completed (${modelToUse})`);
        sendToRenderer('update-status', 'Listening...');
    } catch (error) {
        console.error('[Groq] Error:', error);
        sendToRenderer('update-status', 'Groq error: ' + error.message);
    }
}

/* ── Gemma fallback via Google GenAI ──────────────────────────────────────── */

function trimConversationHistoryForGemma(history, maxChars = 42000) {
    if (!history || history.length === 0) return [];
    let totalChars = 0;
    const trimmed = [];

    for (let i = history.length - 1; i >= 0; i--) {
        const turn = history[i];
        const turnChars = (turn.content || '').length;
        if (totalChars + turnChars > maxChars) break;
        totalChars += turnChars;
        trimmed.unshift(turn);
    }
    return trimmed;
}

async function sendToGemma(transcription) {
    const apiKey = getGeminiApiKey();
    if (!apiKey) {
        console.log('[Gemma] No Gemini API key, skipping');
        return;
    }
    if (!transcription || transcription.trim() === '') return;

    console.log('[Gemma] Sending:', transcription.substring(0, 100) + '...');

    groqConversationHistory.push({ role: 'user', content: transcription.trim() });
    const trimmedHistory = trimConversationHistoryForGemma(groqConversationHistory, 42000);

    try {
        const { GoogleGenAI } = require('@google/genai');
        const ai = new GoogleGenAI({ apiKey });

        const messages = trimmedHistory.map((msg) => ({
            role: msg.role === 'assistant' ? 'model' : 'user',
            parts: [{ text: msg.content }],
        }));

        const messagesWithSystem = [
            { role: 'user', parts: [{ text: currentSystemPrompt }] },
            { role: 'model', parts: [{ text: 'Understood. I will follow these instructions.' }] },
            ...messages,
        ];

        const response = await ai.models.generateContentStream({
            model: 'gemma-3-27b-it',
            contents: messagesWithSystem,
        });

        let fullText = '';
        let isFirst = true;

        for await (const chunk of response) {
            const chunkText = chunk.text;
            if (chunkText) {
                fullText += chunkText;
                sendToRenderer(isFirst ? 'new-response' : 'update-response', fullText);
                isFirst = false;
            }
        }

        if (fullText.trim()) {
            groqConversationHistory.push({ role: 'assistant', content: fullText.trim() });
            if (groqConversationHistory.length > 40) {
                groqConversationHistory = groqConversationHistory.slice(-40);
            }
        }

        sendToRenderer('response-complete', {
            transcription: transcription.trim(),
            response: fullText.trim(),
            model: 'gemma-3-27b-it',
            provider: 'gemma',
        });

        console.log('[Gemma] Response completed');
        sendToRenderer('update-status', 'Listening...');
    } catch (error) {
        console.error('[Gemma] Error:', error);
        sendToRenderer('update-status', 'Gemma error: ' + error.message);
    }
}

/* ── Image analysis via Gemini HTTP ───────────────────────────────────────── */

async function sendImageToGeminiHttp(base64Data, prompt) {
    const apiKey = getGeminiApiKey();
    if (!apiKey) return { success: false, error: 'No Gemini API key' };

    try {
        const { GoogleGenAI } = require('@google/genai');
        const ai = new GoogleGenAI({ apiKey });

        const contents = [
            { inlineData: { mimeType: 'image/jpeg', data: base64Data } },
            { text: prompt },
        ];

        const response = await ai.models.generateContentStream({
            model: 'gemini-2.5-flash',
            contents,
        });

        let fullText = '';
        let isFirst = true;

        for await (const chunk of response) {
            const ct = chunk.text;
            if (ct) {
                fullText += ct;
                sendToRenderer(isFirst ? 'new-response' : 'update-response', fullText);
                isFirst = false;
            }
        }

        console.log('[Gemini HTTP] Image response completed');
        return { success: true, text: fullText };
    } catch (error) {
        console.error('[Gemini HTTP] Error:', error);
        return { success: false, error: error.message };
    }
}

/* ── Public API ────────────────────────────────────────────────────────────── */

function setSystemPrompt(prompt) {
    currentSystemPrompt = prompt || 'You are a helpful assistant.';
}

function getConversationHistory() {
    return groqConversationHistory;
}

function resetConversationHistory() {
    groqConversationHistory = [];
}

/**
 * Generate a text response – picks Groq (primary) or Gemma (fallback).
 */
async function generateTextResponse(transcription) {
    if (hasGroqKey()) {
        await sendToGroq(transcription);
    } else {
        await sendToGemma(transcription);
    }
}

/**
 * Build a context restoration message from conversation history.
 */
function buildContextMessage() {
    const lastTurns = groqConversationHistory.slice(-20);
    // filter paired user/assistant turns
    const pairs = [];
    for (let i = 0; i < lastTurns.length - 1; i++) {
        if (lastTurns[i].role === 'user' && lastTurns[i + 1].role === 'assistant') {
            pairs.push({ user: lastTurns[i].content, assistant: lastTurns[i + 1].content });
        }
    }
    if (pairs.length === 0) return null;

    const lines = pairs.map((p) => `[Interviewer]: ${p.user}\n[Your answer]: ${p.assistant}`);
    return `Session reconnected. Here's the conversation so far:\n\n${lines.join('\n\n')}\n\nContinue from here.`;
}

module.exports = {
    sendToGroq,
    sendToGemma,
    sendImageToGeminiHttp,
    generateTextResponse,
    hasGroqKey,
    getGroqApiKey,
    getGeminiApiKey,
    setSystemPrompt,
    getConversationHistory,
    resetConversationHistory,
    buildContextMessage,
    sendToRenderer,
    getModelForToday,
};

File: stealthpilot-ai-interview-assistant-backend/app/services/groq_service.py
"""
Groq API Service for fast text response generation.

Primary text-response path: Groq with daily model rotation + streaming.
Fallback: Gemma-3-27b-it via Google GenAI.

Mirrors the cheating-daddy architecture where Gemini handles audio
transcription and Groq handles the fast text answer generation.
"""
import asyncio
import logging
from typing import Optional, Dict, Any, AsyncIterator, List

from app.config import get_settings

settings = get_settings()
logger = logging.getLogger(__name__)

# ── Groq daily model rotation ──────────────────────────────────────────────

GROQ_MODELS = [
    "qwen-qwq-32b",
    "deepseek-r1-distill-llama-70b",
    "llama-3.3-70b-versatile",
    "llama-3.1-8b-instant",
    "meta-llama/llama-4-scout-17b-16e-instruct",
]


def _get_model_for_today() -> str:
    """Rotate models daily so rate limits are shared."""
    from datetime import datetime

    day_of_year = datetime.utcnow().timetuple().tm_yday
    return GROQ_MODELS[day_of_year % len(GROQ_MODELS)]


def _strip_thinking_tags(text: str) -> str:
    import re
    return re.sub(r"<think>[\s\S]*?</think>", "", text).strip()


# ── Groq Service ───────────────────────────────────────────────────────────

class GroqService:
    """Async Groq API client with streaming SSE and daily model rotation."""

    def __init__(self):
        self._ready = False
        self._init_error: Optional[str] = None
        self._try_init()

    def _try_init(self):
        api_key = getattr(settings, "GROQ_API_KEY", "")
        if not api_key or api_key == "your-groq-api-key":
            self._init_error = "GROQ_API_KEY not configured in .env"
            logger.warning(f"⚠️  {self._init_error}")
            return
        self._ready = True
        logger.info(f"✅ Groq service ready (model today: {_get_model_for_today()})")

    @property
    def ready(self) -> bool:
        return self._ready

    async def generate_answer(
        self,
        question: str,
        context: Optional[Dict[str, Any]] = None,
        system_prompt: str = "",
        conversation_history: Optional[List[Dict[str, str]]] = None,
    ) -> str:
        """
        Generate an answer via Groq streaming, strip <think> tags, return full text.
        Falls back to a simple error string if the key is missing.
        """
        if not self._ready:
            return f"❌ Groq not available: {self._init_error}"

        import httpx

        api_key = settings.GROQ_API_KEY
        model = _get_model_for_today()

        # Build messages array
        messages: List[Dict[str, str]] = []
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        else:
            context_str = ""
            if context:
                if context.get("company_name"):
                    context_str += f"Company: {context['company_name']}\n"
                if context.get("role_title"):
                    context_str += f"Role: {context['role_title']}\n"
            messages.append({
                "role": "system",
                "content": f"You are an expert interview assistant.{chr(10) + context_str if context_str else ''}"
            })

        # Append conversation history for multi-turn context
        if conversation_history:
            messages.extend(conversation_history[-20:])

        messages.append({"role": "user", "content": question.strip()})

        try:
            full_text = ""
            async with httpx.AsyncClient(timeout=30.0) as client:
                async with client.stream(
                    "POST",
                    "https://api.groq.com/openai/v1/chat/completions",
                    headers={
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json",
                    },
                    json={
                        "model": model,
                        "messages": messages,
                        "stream": True,
                        "temperature": 0.7,
                        "max_tokens": 1024,
                    },
                ) as resp:
                    if resp.status_code != 200:
                        body = await resp.aread()
                        logger.error(f"Groq API error {resp.status_code}: {body.decode()}")
                        return f"Groq error: HTTP {resp.status_code}"

                    async for line in resp.aiter_lines():
                        if not line.startswith("data: "):
                            continue
                        data = line[6:]
                        if data == "[DONE]":
                            break
                        try:
                            import json
                            payload = json.loads(data)
                            token = payload.get("choices", [{}])[0].get("delta", {}).get("content", "")
                            if token:
                                full_text += token
                        except Exception:
                            pass

            cleaned = _strip_thinking_tags(full_text)
            logger.info(f"Groq response completed ({model}): {cleaned[:80]}")
            return cleaned

        except Exception as e:
            logger.error(f"Groq error: {e}")
            return f"Error: {e}"


# ── Gemma fallback ─────────────────────────────────────────────────────────

class GemmaService:
    """Fallback text generation via Gemma-3-27b-it through Google GenAI SDK."""

    def __init__(self):
        self._ready = False
        self._init_error: Optional[str] = None
        self._try_init()

    def _try_init(self):
        if not settings.GEMINI_API_KEY or settings.GEMINI_API_KEY == "your-gemini-api-key":
            self._init_error = "GEMINI_API_KEY not configured"
            return
        self._ready = True
        logger.info("✅ Gemma fallback service ready (gemma-3-27b-it)")

    @property
    def ready(self) -> bool:
        return self._ready

    async def generate_answer(
        self,
        question: str,
        context: Optional[Dict[str, Any]] = None,
        system_prompt: str = "",
        conversation_history: Optional[List[Dict[str, str]]] = None,
    ) -> str:
        if not self._ready:
            return f"❌ Gemma not available: {self._init_error}"

        try:
            import google.generativeai as genai

            genai.configure(api_key=settings.GEMINI_API_KEY)
            model = genai.GenerativeModel("gemma-3-27b-it")

            # Build conversation
            messages = []
            if system_prompt:
                messages.append({"role": "user", "parts": [{"text": system_prompt}]})
                messages.append({"role": "model", "parts": [{"text": "Understood. I will follow these instructions."}]})

            if conversation_history:
                for msg in conversation_history[-20:]:
                    role = "model" if msg["role"] == "assistant" else "user"
                    messages.append({"role": role, "parts": [{"text": msg["content"]}]})

            messages.append({"role": "user", "parts": [{"text": question.strip()}]})

            response = await asyncio.to_thread(
                model.generate_content,
                messages,
            )

            return response.text.strip()

        except Exception as e:
            logger.error(f"Gemma error: {e}")
            return f"Error: {e}"


# ── Singleton instances ────────────────────────────────────────────────────

groq_service = GroqService()
gemma_service = GemmaService()


async def generate_text_response(
    question: str,
    context: Optional[Dict[str, Any]] = None,
    system_prompt: str = "",
    conversation_history: Optional[List[Dict[str, str]]] = None,
) -> str:
    """Primary entry point: try Groq first, fall back to Gemma."""
    if groq_service.ready:
        return await groq_service.generate_answer(question, context, system_prompt, conversation_history)
    if gemma_service.ready:
        return await gemma_service.generate_answer(question, context, system_prompt, conversation_history)
    # Final fallback to existing Gemini text service
    from app.services.gemini_text import gemini_text_service
    return await gemini_text_service.generate_answer(question, context)
